src/tcp_wrap.cc

```diff
@@ -181,6 +183,16 @@ TCPWrap::TCPWrap(Environment* env, Local<Object> object, ProviderType provider)
                    // Suggestion: uv_tcp_init() returns void.
 }
 
+void TCPWrap::Reinitialize(const FunctionCallbackInfo<Value>& args) {
+  TCPWrap* wrap;
+
+  ASSIGN_OR_RETURN_UNWRAP(
+      &wrap, args.Holder(), args.GetReturnValue().Set(UV_EBADF));
+
+  Environment* env = wrap->env();
+  int r = uv_tcp_init(env->event_loop(), &wrap->handle_);
```

Comment:
Should this have some sort of check like this?

```suggestion
  CHECK(!IsAlive(this));
  int r = uv_tcp_init(env->event_loop(), &wrap->handle_);
```

Reply:
Yes, probably. I'll add this.

Reply:
@addaleax I tried, and it always fails, both in `TCPWrap::TCPWrap` or in `TCPWrap::Reinitialize`. So I guess this is not needed. Or am I doing the wrong check?

Reply:
Does that mean that the handle is not closed at this point? I don’t think it’s safe to call `uv_..._init()` functions on open handles? @bnoordhuis

Reply:
The handle just received `uv_tcp_init` but never any open/connect methods. Does it still mean they are considered opened?

Reply:
I think so, yes. If the handle only ever received a `uv_tcp_init` call, why do we need to re-initialize it?

Reply:
I'm not sure if I got this right. Here's my scenario, so feel free to tell me I'm an idiot.

In the context of autoSelectFamily algorithm, the socket attempts a connection to the first address. Let's say a IPv6 one. So internally `TCPWrap::TCPWrap` is called and therefore the `uv_tcp_init` and then `uv_tcp_connect`.

Now, if the connection fails, the algorithm will try to connect to the second address, let's say a IPv4, using the same TCPWrap. In this case I thought we need to reset the handle and therefore I introduced `Reinitialize` which will call `uv_tcp_init` again before the second `uv_tcp_connect`.

Or can be `uv_tcp_connect` called on the same handle with no issues?

Reply:
> using the same TCPWrap. 

Just to say, you might have an easier time here by using separate TCPWrap instances.

> In this case I thought we need to reset the handle and therefore I introduced Reinitialize which will call uv_tcp_init again before the second uv_tcp_connect.

Yeah, I’m pretty sure that’s not allowed by libuv. `uv_tcp_init` is not idempotent.

> Or can be uv_tcp_connect called on the same handle with no issues?

As @RaisinTen said, I think this is worth trying. I assume the original connect request will have already been completed by the time we re-try this.

Reply:
Shouldn't we call [`uv_close()`](http://docs.libuv.org/en/v1.x/handle.html#c.uv_close) on the handle before calling `uv_tcp_init()` again? Otherwise, the memory allocated by the previous `uv_tcp_init()` call would probably leak.

Reply:
@RaisinTen 🤦
Now that you write it, it feels so obvious. Sob.

I'll make the change soon. Thanks!

Reply:
Yeah, you’ll want at least `uv_close()` (probably via `Close()`?), the `CHECK(!IsAlive(this))` mentioned above, you’ll want to call `MarkAsInitialized()` and `MarkAsUninitialized()` as needed to ensure the current `TCPWrap` object has the right state.

Reply:
@addaleax I tried with `CHECK(!wrap->IsAlive())` but in my situation always fails. Therefore, I'm not messing with `MarkAsInitialized()` and `MarkAsUninitialized()`. Therefore I used a slightly different approach by calling `Close` only if `uv_is_active` returns `true`.

How does it look now?

Reply:
I mean … unfortunately, it doesn’t really look different, because it’s invalid to re-initialize a handle regardless of whether it is active or not?

Reply:
I think I'm blocked here then. Do you mind taking a look to the branch and try to make the changes yourself?
In order to test them, run them against all the test in `test/parallel` containing `autoselectfamily` in the name.

Reply:
I’m not sure when exactly I’ll have time to do that – I did take another look at the code here and it’s clear that this can’t work, `Close()` is an async operation (because the underlying `uv_close()` is), so you can’t call `InitializeHandle()` _immediately_ afterwards, you need to wait for the close operation to finish first

Reply:
`handle_` can’t be accessed directly from JS. I don’t know how relevant that is anyway, though, because even if it were, I don’t think we would consider userland accesses to it allowed.

> The reason I'm asking is because being `uv_close` async might be a problem since `lib/net.js` code leading to `uv_tcp_connect` is a sync method.

It probably makes sense to make the JS code here async as well. I don’t think there’s anything speaking against that, there aren’t any synchronous public APIs involved here, right?

Reply:
Although tbh – yes, there is some trickiness here. If there is a period during which the handle is closing before it is re-initialized, the public methods on the socket should still work if they are called (`.ref()`, `.unref()`, `.remoteAddress`, `.localAddress`, etc.)

Reply:
Exactly.
I was leaning to the idea, when `Reinitialize` is called, to call `uv_close` on the already initialized `handle_` and, at the same time, creating a new `handle_` in the object so that the method you mentioned above refer to the newer handle.

I'm not familiar too much with C++ memory handling, so I ask you: is this approach viable?

In other words, in C++ pseudo code:

```
void TCPWrap::Reinitialize() {
  uv_close(handle_, NULL) // No callback needed as in theory we don't care what happens to the old object
  this->handle_ = create_handle_somehow() // Sorry, on iPad at the moment so I can't really access source code
  uv_tcp_connect(this->handle_);
}
```

Do you think this might work or there are huge problems I'm not seeing?

Reply:
The pointer to the `uv_tcp_t` object needs to stay alive until `uv_close()` is done with it. At the very least you'd need to pass a callback that frees the memory afterwards.

Reply:
That is fine. I can definetely do that.

Memory problems aside, do you foresee any problems in swapping the handler?

Reply:
I think in theory this works, but it’s going to be a bit more complex than the code above.

For example, you’ll need to update all `ConnectionWrap` subclasses to have a pointer handle member rather than a plain member, and you’re going to have to use `Environment::CloseHandle()` rather than `uv_close()` so that the close callback gets tracked properly.

My gut feeling would be that at that point, it probably makes more sense to look into how to handle this more properly on the JS layer rather than trying to re-use `TCPWrap` instances.

Reply:
Ok, you are probably right, I'm looking for too much troubles. I'll see how to fix it on the JS side.

Reply:
@addaleax I finally solved (hopefully all) the issues without touching the C++ and swapping handles.
Thanks for not having me kill with my own hands :)